src/random_ray/random_ray.cpp

```diff
@@ -174,6 +177,53 @@ float exponentialG2(float tau)
   return num / den;
 }
 
+// Implementation of the Fisher-Yates shuffle algorithm.
+// Algorithm adapted from:
+//    https://en.cppreference.com/w/cpp/algorithm/random_shuffle#Version_3
+void fisher_yates_shuffle(vector<int64_t>& arr, uint64_t* seed)
+{
+  // Loop over the array from the last element down to the second
+  for (int i = arr.size() - 1; i > 0; --i) {
+    // Generate a random index in the range [0, i]
+    int j = uniform_int_distribution(0, i, seed);
+    std::swap(arr[i], arr[j]);
+  }
+}
+
+// Function to generate randomized Halton sequence samples
+//
+// Algorithm adapted from:
+//      A. B. Owen. A randomized halton algorithm in r. Arxiv, 6 2017.
+//      URL https://arxiv.org/abs/1706.02808
+vector<double> rhalton(int dim, uint64_t* seed, int64_t skip = 0)
+{
+  int64_t b, res, dig;
+  double b2r, ans;
+  vector<int64_t> primes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
```

Comment:
```suggestion
const std::array<int64_t, 10> primes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
```
This could be faster, as it avoids having to dynamically allocate memory for the vector. The `std::array` is statically sized and is on the stack, so probably is faster. Not sure how much of a difference it would make in practice, but could be useful in some cases potentially, e.g., rejection sampling, where the ray sampling scheme needs to be fast.